crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! Its goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which will be used by this allocator is a single page of memory, which in Wasm is
+//! 64KiB. We do not expect contracts to use more memory than this (for now), so we will throw an
+//! OOM error instead of requesting more memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is 64KiB
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+/// A bump allocator suitable for use in a Wasm environment.
+pub struct BumpAllocator;
+
+impl BumpAllocator {
+    /// Initialize the backing heap of the bump allocator.
+    ///
+    /// This function must only be called **once**, and it **must** be called before any
+    /// allocations are made.
+    #[inline]
+    pub fn init(&self) {
+        // SAFETY: We are in a single threaded context, so we don't have to worry about this being
+        // concurrently mutated by multiple threads.
+        unsafe { INNER.init() }
+    }
+}
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        INNER.alloc(layout)
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+struct InnerAlloc {
+    /// Points to the start of the next available allocation.
+    next: usize,
+
+    /// The address of the upper limit of our heap.
+    upper_limit: usize,
+}
+
+impl InnerAlloc {
+    const fn new() -> Self {
+        Self {
+            next: 0,
+            upper_limit: 0,
+        }
+    }
+
+    /// Initialize the heap which backs the bump allocator.
+    ///
+    /// Our heap is a single page of Wasm memory (64KiB) and will not grow beyond that.
+    ///
+    /// Note that this function must be called before any allocations can take place, otherwise any
+    /// attempts to perform an allocation will fail.
+    fn init(&mut self) {
+        let prev_page = core::arch::wasm32::memory_grow(0, 1);
+        if prev_page == usize::MAX {
+            todo!()
```

Comment:
What we do instead of these `todo!`'s will depend on the context in which `init()` is called

Reply:
The `init` function could return a `enum InitState { AlreadyInitialized, NowInitialized }` value to handle those `todo!` cases.

Reply:
In the case of `prev_page == usize::MAX` we should panic since that's really an exceptional case where we can do nothing.